include/RAJA/index/ListSegment.hpp

```diff
@@ -165,20 +175,102 @@ class TypedListSegment
+  ///        and use given camp resource to allocate list segment index data
+  ///        if owned by this list segment.
+  ///
+  /// By default the ctor performs a deep copy of array elements.
+  ///
+  /// If 'Unowned' is passed as last argument, the constructed object
+  /// does not own the segment data and will hold a pointer to given
+  /// array's data. In this case, caller must manage object lifetimes properly.
+  ///
+  TypedListSegment(const value_type* values,
+                   Index_type length,
+                   camp::resources::Resource& resource,
+                   IndexOwnership owned = Owned)
+    : m_resource(resource), m_use_resource(true)
+  {
+    initIndexData(m_use_resource,
+                  values, length, owned);
+  }
+
+  ///
+  /// Construct list segment from arbitrary object holding
+  /// indices using a deep copy of given data.
+  ///
+  /// The object must provide methods: begin(), end(), size().
+  ///
+  template <typename Container>
+  TypedListSegment(const Container& container,
+                   camp::resources::Resource& resource)
+    : m_resource(resource), m_use_resource(true),
+      m_owned(Unowned), m_data(nullptr), m_size(container.size())
+  {
+    using namespace camp::resources;
+
+    if (m_size > 0) {
+
+      Resource host_res{Host()};
+
+      value_type* tmp = host_res.allocate<value_type>(m_size);
+
+      auto dest = tmp;
+      auto src = container.begin();
+      auto const end = container.end();
+      while (src != end) {
+        *dest = *src;
+        ++dest;
+        ++src;
+      }
+
+      m_data = m_resource.allocate<value_type>(m_size);
+      m_resource.memcpy(m_data, tmp, sizeof(value_type) * m_size);
+      m_owned = Owned;
+
+      host_res.deallocate(tmp);
+
+    }
+  }
+
+
+/*
+ * The following two ctors preserve the original list segment behavior for
+ * CUDA and HIP device memory management. 
+ *
+ * Note that the host resource object created in the member initialization
+ * list is not used. Where memory management routines are shared between
+ * the old way and using camp resources are controlled by the m_use_resource
+ * boolean member.
+ */
+
   ///
   /// \brief Construct list segment from given array with specified length.
   ///
   /// By default the ctor performs deep copy of array elements.
+  ///
   /// If 'Unowned' is passed as last argument, the constructed object
-  /// does not own the segment data and will hold a pointer to given data.
-  /// In this case, caller must manage object lifetimes properly.
+  /// does not own the segment data and will hold a pointer to given
+  /// array's data. In this case, caller must manage object lifetimes properly.
   ///
   TypedListSegment(const value_type* values,
                    Index_type length,
                    IndexOwnership owned = Owned)
+    : m_resource(camp::resources::Resource{camp::resources::Host()}),
+      m_use_resource(false)
```

Comment:
Should default init all members where possible

Reply:
Done!